pkg/protoassert/assert.go

```diff
@@ -1,130 +1,194 @@
+// Package protoassert provides utility functions for comparing protobuf-generated objects in tests.
+//
+// These functions are based on related ones found in https://github.com/stretchr/testify.
 package protoassert
 
 import (
+	"bytes"
+	"fmt"
 	"testing"
 
+	"github.com/davecgh/go-spew/spew"
 	"github.com/stretchr/testify/assert"
-	"github.com/stretchr/testify/require"
 	"golang.org/x/exp/maps"
-	"google.golang.org/protobuf/encoding/protojson"
 	"google.golang.org/protobuf/proto"
 )
 
 type message[V any] interface {
 	*V
 	proto.Message
-	String() string
-	EqualVT(other *V) bool
+	fmt.Stringer
+	EqualVT(that *V) bool
 }
 
-func Equal[V any, T message[V]](t testing.TB, expected, actual T, msgAndArgs ...interface{}) bool {
+var spewConfig = spew.ConfigState{
+	Indent:                  " ",
+	DisablePointerAddresses: true,
+	DisableCapacities:       true,
+	SortKeys:                true,
+	DisableMethods:          true,
+	MaxDepth:                10,
+}
+
+// Equal mimics [assert.Equal].
+func Equal[T message[V], V any](t testing.TB, expected, actual T, msgAndArgs ...any) bool {
 	t.Helper()
-	if expected.EqualVT(actual) {
-		return true
+	if !actual.EqualVT(expected) {
+		return assert.Fail(t, fmt.Sprintf("Not equal: \n"+
+			"expected: %s\n"+
+			"actual  : %s", expected, actual), msgAndArgs...)
 	}
-	e, err := toJson(expected)
-	require.NoError(t, err)
-	a, err := toJson(actual)
-	require.NoError(t, err)
-	return assert.JSONEq(t, e, a, msgAndArgs)
+	return true
 }
 
-func NotEqual[V any, T message[V]](t testing.TB, expected, actual T, msgAndArgs ...interface{}) bool {
+// NotEqual mimics [assert.NotEqual].
+func NotEqual[T message[V], V any](t testing.TB, expected, actual T, msgAndArgs ...any) bool {
 	t.Helper()
-	return assert.False(t, expected.EqualVT(actual), msgAndArgs)
+	if actual.EqualVT(expected) {
+		return assert.Fail(t, fmt.Sprintf("Should not be: %#v\n", actual), msgAndArgs...)
```

Comment:
The problem with `%s` is that it misses `nil` values, which typically is ok, but for the original purpose of this PR, it's paramount. For the protos, `%v` and `%+v` have the same problem. At least with `%#v` we can see the types. In my particular example I can see

Expected: `HasLayerIndex:storage.isEmbeddedImageScanComponent_HasLayerIndex(nil)`
Actual: `HasLayerIndex:(*storage.EmbeddedImageScanComponent_LayerIndex)(nil)`

This will be masked with the other options.

I tried looking into how to remove unexported fields, and I can't find anything. I do understand it's noisy, but I think keeping `nil` and type information is worth the noise